### ISSUE TITLE
Support Props in the Router

### ISSUE THREAD

**Issue by Mahoney** [japgolly/scalajs-react#424]
I keep finding cases where I want to have State that lives above a Router.

For instance:
```
StateHoldingComponent(state: State)
  |-Router
      |-StateEditComponent(props: State, stateUpdater: StateRW)
      |-StateVisualisationComponent(props: State)
```
where I want the Router functionality to make bookmarkable URLs, but changes to the state in the StateEditComponent should obviously be visible in the StateVisualisationComponent. I still want to build the Router once, but construct it with some props in the usual way to pass the State down and make it available to the child components as so:

```scala
(emptyRule
      | staticRoute("edit", Edit)           ~> { props => StateEditComponent(props.state) }
      | staticRoute("visualise", Visualise) ~> { props => StateVisualisationComponent(props.state) }
```

Another use case is wanting to do a redirect to a default URL for the current user:
```scala
(emptyRule
      | staticRoute(root, Root) ~> { props => props.maybeDefaultPage.fold(
                                       render(NoDefaultPage(props.user)): Action
                                     )( defaultPage =>
                                       redirectToPage(Page(defaultPage))(Redirect.Replace): Action
                                     )
                                   }
```

To facilitate this, I'd like the Router to optionally take some props and make them available to the function passed to ~>.

I'm having a go at adding this feature to the Router, but finding the level of FP in the code a challenge. If I put up a probably rather flawed attempt at it, would it be of interest?

**Comment by japgolly** [comment]
I've got a similar situation in my own project where there's top-level state for the SPA and it even affects routes. I came up with a different approach than you envision; philosophically I think it's very important that the top-most component be the router, simply because it serves as a compile-proof that the router is always on, always active, that your app never stops being an SPA. If you have a different top-level component that calls the router with props, then if also has the power to simply not call the router at all. Now obviously you'd never do that (right?) but I my thoughts were that having the Router at the top-level really solidifies architecturally how SPA's work, and are expected to work.

I've created a gist here with excerpts from my main project and sample code to demonstrate my approach: https://gist.github.com/japgolly/7026294793e3e49030b126cd4ff521af

WDYT?

**Comment by Mahoney** [comment]
That's nice... I'll spike out moving my app to that structure and see how it works.

**Comment by japgolly** [comment]
Hi @Mahoney, did you try out this approach, and did it solve your problem?

**Comment by Mahoney** [comment]
Hi - apologies for the radio silence. It worked fine for my first use case, but I failed to get it to work for my second use case because I couldn't find any way to do the redirect. The ```LoadedRoot .Backend.render``` method has (as far as I can see) to return a vdomelement, so it can't redirect the router.

**Comment by japgolly** [comment]
Hey @Mahoney - no apologies necessary! :)

> The LoadedRoot .Backend.render method has (as far as I can see) to return a vdomelement, so it can't redirect the router.

It should be the job of the routing rules to determine whether to redirect or render, trying to squash that in to `Backend.render` isn't going to be pretty. Have you tried moving that logic into the routing rules?

**Comment by japgolly** [comment]
Hmmm, maybe [this](https://github.com/japgolly/scalajs-react/blob/master/extra/src/main/scala/japgolly/scalajs/react/extra/router/RouterConfig.scala#L125) should be wrapped in `CallbackTo`. The current type signature implies that route actions are always pure.

Alternatively you could use [`.addCondition`](https://github.com/japgolly/scalajs-react/blob/master/extra/src/main/scala/japgolly/scalajs/react/extra/router/Dsl.scala#L369-L392) which takes a `CallbackTo` which can be derived from state in a type-safe manner.

**Comment by japgolly** [comment]
It's not optimal, but you could also just weaken your type-safety. Inside your `dynRender`/`dynRenderR` method you could call `.runNow()` on your state. I can't promise there won't be any bugs, I don't remember the details of how everything in that space works (which is the whole point of `CallbackTo` allowing scalac to prevent them) but it'll probably be fine. It's probably either that or a PR for `RouterConfig.Parsed`.

**Comment by japgolly** [comment]
Closed by #682
